### PR TITLE
[SYCLomatic] Migrate thrust::gather_if

### PR DESCRIPTION
**Pull request by LU-JOHN**
Signed-off-by: Lu, John <john.lu@intel.com>

### CODE REVIEW

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/test/dpct/thrust_gather_usm.cu [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+// CHECK-NEXT: thrust::gather_if(AD.begin(), AD.end(), SD.begin(), BD.begin(), RD.begin());
+// CHECK-NEXT: /*
+// CHECK-NEXT: DPCT1107:{{[0-9]+}}: Migration for this overload of thrust::gather_if is not supported.
+// CHECK-NEXT: */
+// CHECK-NEXT: thrust::gather_if(h_ptr, h_ptr + 4, SH.begin(), BH.begin(), RH.begin());
+  // VERSION                        first       last      stencil     input       result
+  thrust::gather_if(                AH.begin(), AH.end(), SH.begin(), BH.begin(), RH.begin());
+  thrust::gather_if(                AD.begin(), AD.end(), SD.begin(), BD.begin(), RD.begin());
+  thrust::gather_if(                h_ptr,      h_ptr+4,  SH.begin(), BH.begin(), RH.begin());
+#ifdef ADD_BUG
```
Comment (0x12CC) [inline clang/test/dpct/thrust_gather_usm.cu]: Why is this needed? Can it be removed?

Comment (LU-JOHN) [reply clang/test/dpct/thrust_gather_usm.cu]: nvcc does not accept this overload, but I believe this is an oversight in nvcc.  I believe it is good to have a record of this so if nvcc is ever fixed we can quickly add testing for this.

File: clang/test/dpct/test_api_level/check_api_level_test_coverage.cu [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 
 // CHECK: begin
+// CHECK-NEXT: DplExtrasAlgorithm_gather_if
```
Comment (jzc) [inline clang/test/dpct/test_api_level/check_api_level_test_coverage.cu]: You need to a corresponding api_test for `DplExtrasAlgorithm_gather_if` in the `test_api_level` directory instead of adding a `CHECK` here.

**Review by jzc — CHANGES_REQUESTED** [review]

File: clang/test/dpct/test_api_level/check_api_level_test_coverage.cu (hunk 2 above)
Comment (LU-JOHN) [reply clang/test/dpct/test_api_level/check_api_level_test_coverage.cu]: Test in test_api_level directory added.

File: clang/test/dpct/thrust_gather_usm.cu (hunk 1 above)
Comment (LU-JOHN) [reply clang/test/dpct/thrust_gather_usm.cu]: Failing code added commented out with #ifdef.

**Review by jzc — APPROVED** [review]

**Review by 0x12CC — APPROVED** [review]
